Kernel/Devices/EBRPartitionTable.cpp

```diff
@@ -0,0 +1,203 @@
+}
+
+const EBRPartitionExtension& EBRPartitionTable::ebr_extension() const
+{
+    return *reinterpret_cast<const EBRPartitionExtension*>(m_cached_ebr_header);
+}
+
+int EBRPartitionTable::get_ebr_container() const
+{
+    for (int i = 0; i < 4; i++)
+        if (header().entry[i].type == EBR_CHS_CONTAINER || header().entry[i].type == EBR_LBA_CONTAINER)
+            return i;
+    ASSERT_NOT_REACHED();
+}
+
+bool EBRPartitionTable::initialize()
+{
+    if (!m_device->read_block(0, m_cached_mbr_header)) {
+        return false;
+    }
+    auto& header = this->header();
+
+    m_ebr_container_id = get_ebr_container() + 1;
+
+#ifdef EBR_DEBUG
+    kprintf("EBRPartitionTable::initialize: MBR_signature=%#x\n", header.mbr_signature);
+#endif
+
+    if (header.mbr_signature != MBR_SIGNATURE) {
+        kprintf("EBRPartitionTable::initialize: bad MBR signature %#x\n", header.mbr_signature);
+        return false;
+    }
+
+    auto& ebr_entry = header.entry[m_ebr_container_id - 1];
+    size_t index = 1;
+    if (!m_device->read_block(ebr_entry.offset, m_cached_ebr_header)) {
+        return false;
+    }
+    while (index < 128) { // Unlikely to encounter a disk with 128 partitions in this configuration...
+        if (ebr_extension().next_chained_ebr_extension.offset == 0 && ebr_extension().next_chained_ebr_extension.type == 0) {
+            break;
+        }
+        index++;
+        if (!m_device->read_block(ebr_extension().next_chained_ebr_extension.offset, m_cached_ebr_header)) {
+            return false;
+        }
+    }
+
+    m_ebr_chained_extensions_count = index;
+
+    kprintf("EBRPartitionTable::initialize: Extended partitions count - %d\n", m_ebr_chained_extensions_count);
+
+    return true;
+}
+
+RefPtr<DiskPartition> EBRPartitionTable::get_non_extended_partition(unsigned index)
+{
+    auto& header = this->header();
+    auto& entry = header.entry[index - 1];
+
+#ifdef EBR_DEBUG
+    kprintf("EBRPartitionTable::partition: status=%#x offset=%#x\n", entry.status, entry.offset);
+#endif
+
+    if (entry.offset == 0x00) {
+#ifdef EBR_DEBUG
+        kprintf("EBRPartitionTable::partition: missing partition requested index=%d\n", index);
+#endif
+
+        return nullptr;
+    }
+
+#ifdef EBR_DEBUG
+    kprintf("EBRPartitionTable::partition: found partition index=%d type=%x\n", index, entry.type);
+#endif
+
+    return DiskPartition::create(m_device, entry.offset, (entry.offset + entry.length));
+}
+
+RefPtr<DiskPartition> EBRPartitionTable::get_extended_partition(unsigned index)
+{
+
+    unsigned relative_index = index - m_ebr_container_id;
+    auto& header = this->header();
+
+#ifdef EBR_DEBUG
+    kprintf("EBRPartitionTable::partition: relative index %d\n", relative_index);
+#endif
+
+    auto& ebr_entry = header.entry[m_ebr_container_id - 1];
+#ifdef EBR_DEBUG
+    kprintf("EBRPartitionTable::partition: Extended partition, offset 0x%x, type %x\n", ebr_entry.offset, ebr_entry.type);
+#endif
+
+    size_t i = 0;
```

Comment:
Since this variable is not used until later, we should declare it as close as possible to it being used.